File: app/src/main/java/com/example/ivdemo/IPCActivity.java
package com.example.ivdemo;

import android.graphics.SurfaceTexture;
import android.os.Bundle;
import android.util.Log;
import android.view.TextureView;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;
import android.content.Context;
import com.tencent.iot.voipdemo.R;

import androidx.appcompat.app.AppCompatActivity;

import com.tencent.iot.voip.device.VoipNativeInterface;
import com.tencent.iot.voip.device.callback.IvAvtCallback;
import com.tencent.iot.voip.device.callback.IvDeviceCallback;
import com.tencent.iot.voip.device.consts.CommandType;
import com.tencent.iot.voip.device.consts.P2pEventType;
import com.tencent.iot.voip.device.consts.StreamType;
import com.tencent.iot.voip.device.model.AvtInitInfo;
import com.tencent.iot.voip.device.model.DeviceInfo;
import com.tencent.iot.voip.device.model.SysInitInfo;
import com.tencent.iotvideo.link.CameraRecorder;


public class IPCActivity extends AppCompatActivity implements IvAvtCallback ***REMOVED***
    private static final String TAG = IPCActivity.class.getSimpleName();

    protected SimplePlayer mPlayer;
    protected CameraRecorder mCameraRecorder;
    // view for remote video
    protected TextureView mRemoteView;
    // view for local camera preview
    protected SurfaceTexture mLocalPreviewSurface;
    // view for remote preview
    protected SurfaceTexture mRemotePreviewSurface;
    protected TextureView mTextureView;
    // display device info
    protected TextView mTextDevinfo;
    protected String mProductId;
    protected String mDeviceName;

    private static long lastClickTime;

    protected void initWidget() ***REMOVED***
        setContentView(R.layout.activity_ipc);
//        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT)

        mTextDevinfo = findViewById(R.id.text_ipc_devinfo);
        // Find the TextureView in the layout
        mTextureView = findViewById(R.id.textureView_ipc);

        mPlayer = new SimplePlayer();
        mCameraRecorder = new CameraRecorder();

        Button callButton = findViewById(R.id.btn_ipc_call);
        callButton.setOnClickListener(new View.OnClickListener() ***REMOVED***
            @Override
            public void onClick(View v) ***REMOVED***
                long time = System.currentTimeMillis();
                long timeD = time - lastClickTime;
                //防止频繁点击
                if (0 < timeD && timeD < 1000) ***REMOVED***
                    Toast.makeText(IPCActivity.this, "频繁点击！", Toast.LENGTH_SHORT).show();
                    return;
              ***REMOVED***
                lastClickTime = time;
                // msg_id 6: 按门铃
//                IVoipJNIBridge.getInstance().sendMsgNotice(6);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***


    @Override
    protected void onCreate(Bundle savedInstanceState) ***REMOVED***
        Log.d(TAG, "start create");
        super.onCreate(savedInstanceState);

        // set device info
        mProductId = getIntent().getStringExtra("productId");
        mDeviceName = getIntent().getStringExtra("deviceName");
        String deviceKey = getIntent().getStringExtra("deviceKey");
        String region = "china";
        String devinfo = mProductId + "/" + mDeviceName;

        // local AV files for playback
        AssetFileOps fileOps = new AssetFileOps();
        String path = getFilesDir().getAbsolutePath();
        Log.d(TAG, "path is " + path);
        String audioFileName = "audio_sample16000_stereo_64kbps.aac";
        String absAudioFileName = path + "/" + audioFileName;
        String videoFileName = "video_size320x180_gop50_fps25.h264";
        String absVideoFileName = path + "/" + videoFileName;
        fileOps.copyFileFromAssets(getApplicationContext(), audioFileName, absAudioFileName);
        fileOps.copyFileFromAssets(getApplicationContext(), videoFileName, absVideoFileName);

        // start run JNI iot_video_demo
        SysInitInfo info = SysInitInfo.createDefaultSysInitInfo(new DeviceInfo(mProductId, mDeviceName, deviceKey, region));
        VoipNativeInterface.getInstance().initIvSystem(info, new IvDeviceCallback() ***REMOVED***
            @Override
            public void onOnline(long netDateTime) ***REMOVED***
                Log.d(TAG, "onOnline  netDateTime--->" + netDateTime);
          ***REMOVED***

            @Override
            public void onOffline(int status) ***REMOVED***
                Log.d(TAG, "onOffline  status--->" + status);
          ***REMOVED***

            @Override
            public void onModuleStatus(int moduleStatus) ***REMOVED***
                Log.d(TAG, "moduleStatus--->" + moduleStatus);
          ***REMOVED***
      ***REMOVED***);
        VoipNativeInterface.getInstance().initIvDm();
        AvtInitInfo avtInitInfo = AvtInitInfo.createDefaultAvtInitInfo();
        VoipNativeInterface.getInstance().initIvAvt(avtInitInfo, this);
        Log.d(TAG, "run iot_video_demo for " + devinfo);

        initWidget();
        mTextDevinfo.setText(devinfo);
  ***REMOVED***

    @Override
    protected void onDestroy() ***REMOVED***
        VoipNativeInterface.getInstance().exitWxCloudVoip();
        VoipNativeInterface.getInstance().exitIvAvt();
        VoipNativeInterface.getInstance().exitIvDm();
        VoipNativeInterface.getInstance().exitIvSys();
        super.onDestroy();
  ***REMOVED***

    @Override
    public void onStartRealPlay(int visitor, int channel, int res_type) ***REMOVED***
        Log.d(TAG, "onStartRealPlay  visitor " + visitor + " channel " + channel + " res_type " + res_type);
        mCameraRecorder.startRecording(visitor, res_type);
  ***REMOVED***

    @Override
    public void onStopRealPlay(int visitor, int channel, int res_type) ***REMOVED***
        Log.d(TAG, "onStopRealPlay  visitor " + visitor + " channel " + channel + " res_type " + res_type);
        mCameraRecorder.stopRecording(visitor, res_type);
  ***REMOVED***

    @Override
    public int onStartRecvAudioStream(int visitor, int channel, int type, int option, int mode, int width, int sample_rate, int sample_num) ***REMOVED***
        Log.d(TAG, "onStartRecvAudioStream visitor " + visitor);
        return mPlayer.startAudioPlay(visitor, type, option, mode, width, sample_rate, sample_num);
  ***REMOVED***

    @Override
    public int onStartRecvVideoStream(int visitor, int channel, int type, int height, int width) ***REMOVED***
        Log.w(TAG, "onStartRecvVideoStream  video stream is not supported in this activity");
        return 0;
  ***REMOVED***

    @Override
    public void onNotify(int event, int visitor, int channel, int videoResType) ***REMOVED***
        Log.w(TAG, "onNotify()");
        String msg = "";
        switch (event)***REMOVED***
            case P2pEventType.IV_AVT_EVENT_P2P_PEER_CONNECT_FAIL:
            case P2pEventType.IV_AVT_EVENT_P2P_PEER_ERROR:
            ***REMOVED***
                Log.d(TAG,"receive event: peer error");
                msg = "network err";
          ***REMOVED***
            break;
            case P2pEventType.IV_AVT_EVENT_P2P_PEER_ADDR_CHANGED: ***REMOVED***
                Log.d(TAG,"receive event: peer addr change");
                msg = "peer change";
          ***REMOVED***
            break;
            case P2pEventType.IV_AVT_EVENT_P2P_PEER_READY:
            case P2pEventType.IV_AVT_EVENT_P2P_WATERMARK_LOW:
            case P2pEventType.IV_AVT_EVENT_P2P_WATERMARK_WARN:
            case P2pEventType.IV_AVT_EVENT_P2P_WATERMARK_HIGH:
            case P2pEventType.IV_AVT_EVENT_P2P_LOCAL_NET_READY:
                // do nothing
                break;

            default:
                Log.d(TAG,"not support event");
                msg = "unsupport event type " + event;
                break;
      ***REMOVED***
        if (!msg.isEmpty())***REMOVED***
            updateUI(this, msg);
      ***REMOVED***
  ***REMOVED***

    public int onStopRecvStream(int visitor, int channel, int streamType) ***REMOVED***
        Log.d(TAG, "onStopRecvStream visitor " + visitor + " stream_type " + streamType + " stopped");
        if (streamType == StreamType.IV_AVT_STREAM_TYPE_VIDEO) ***REMOVED***
            return mPlayer.stopVideoPlay(visitor);
      ***REMOVED*** else ***REMOVED***
            return mPlayer.stopAudioPlay(visitor);
      ***REMOVED***
  ***REMOVED***

    @Override
    public int onRecvStream(int visitor, int streamType, byte[] data, int len, long pts, long seq) ***REMOVED***
        Log.d(TAG, "onRecvStream visitor " + visitor + " stream_type " + streamType + " data" + data + "  len" + len + "   pts" + pts + "   seq" + seq);
        if (streamType == 1) ***REMOVED***
            return mPlayer.playVideoStream(visitor, data, len, pts, seq);

      ***REMOVED*** else if (streamType == 0) ***REMOVED***
            return mPlayer.playAudioStream(visitor, data, len, pts, seq);
      ***REMOVED***
        return 0;
  ***REMOVED***

    @Override
    public int onRecvCommand(int command, int visitor, int channel, int videoResType, Object args) ***REMOVED***
        Log.d(TAG, "onRecvCommand command " + command + " visitor " + visitor + " channel" + channel + "   videoResType" + videoResType + "   args" + args);
        String msg = "";
        switch (command) ***REMOVED***
            case CommandType.IV_AVT_COMMAND_USR_DATA: ***REMOVED***
                Log.d(TAG, "receive command: user data");
                msg = "user data";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_REQ_STREAM: ***REMOVED***
                Log.d(TAG, "receive command: request stream");
                msg = "request stream";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CHN_NAME: ***REMOVED***
                Log.d(TAG, "receive command: get channel name");
                msg = "get channel name";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_REQ_IFRAME: ***REMOVED***
                Log.d(TAG, "receive command: request I frame");
                msg = "request I frame";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_PAUSE: ***REMOVED***
                Log.d(TAG, "receive command: playback pause");
                msg = "playback pause";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_RESUME: ***REMOVED***
                Log.d(TAG, "receive command: playback resume");
                msg = "playback resume";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_QUERY_MONTH: ***REMOVED***
                Log.d(TAG, "receive command: playback query month");
                msg = "playback query month";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_QUERY_DAY: ***REMOVED***
                Log.d(TAG, "receive command: playback query day");
                msg = "playback query day";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_SEEK: ***REMOVED***
                Log.d(TAG, "receive command: playback seek");
                msg = "playback seek";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_FF: ***REMOVED***
                Log.d(TAG, "receive command: playback fast forward");
                msg = "playback fast forward";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_SPEED: ***REMOVED***
                Log.d(TAG, "receive command: playback speed");
                msg = "playback speed";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_REWIND: ***REMOVED***
                Log.d(TAG, "receive command: playback rewind");
                msg = "playback rewind";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_PLAYBACK_PROGRESS: ***REMOVED***
                Log.d(TAG, "receive command: playback progress");
                msg = "playback progress";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_QUERY_FILE_LIST: ***REMOVED***
                Log.d(TAG, "receive command: get file list");
                msg = "get file list";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CALL_ANSWER: ***REMOVED***
                Log.d(TAG, "receive command: call answer");
                msg = "call answer";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CALL_HANG_UP: ***REMOVED***
                Log.d(TAG, "receive command: call hang up");
                msg = "call hang up";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CALL_REJECT: ***REMOVED***
                Log.d(TAG, "receive command: call reject");
                msg = "call reject";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CALL_CANCEL: ***REMOVED***
                Log.d(TAG, "receive command: call cancel");
                msg = "call cancel";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CALL_BUSY: ***REMOVED***
                Log.d(TAG, "receive command: call busy");
                msg = "call busy";
          ***REMOVED***
            break;
            case CommandType.IV_AVT_COMMAND_CALL_TIMEOUT: ***REMOVED***
                Log.d(TAG, "receive command: call timeout");
                msg = "call timeout";
          ***REMOVED***
            break;

            default:
                Log.d(TAG, "not support command");
                msg = "unsupport cmd type " + command;
                break;
      ***REMOVED***
        // Toast.makeText(IPCActivity.this, msg, Toast.LENGTH_SHORT).show();
        updateUI(this, msg);
        return 0;
  ***REMOVED***

    @Override
    public int onDownloadFile(int status, int visitor, int channel, Object args) ***REMOVED***
        Log.d(TAG, "onDownloadFile status " + status + " visitor " + visitor + " channel"+channel+"   args"+args);
        return 0;
  ***REMOVED***

    @Override
    public void onGetPeerOuterNet(int visitor, int channel, String netInfo) ***REMOVED***
        Log.d(TAG, "onGetPeerOuterNet visitor " + visitor + " channel " + channel + " netInfo"+netInfo);
  ***REMOVED***

    public void updateUI(Context context, String msg) ***REMOVED***
        ((IPCActivity) context).runOnUiThread(new Runnable() ***REMOVED***
            @Override
            public void run() ***REMOVED***
                //此时已在主线程中，可以更新UI了
                Toast.makeText(IPCActivity.this, msg, Toast.LENGTH_SHORT).show();
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
}

File: app/src/main/java/com/example/ivdemo/IPCActivity.kt
package com.example.ivdemo

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.SurfaceTexture
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.TextureView
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.lifecycle.lifecycleScope
import com.example.ivdemo.popup.CustomCommandDialog
import com.example.ivdemo.popup.QualitySettingDialog
import com.tencent.iot.twcall.R
import com.tencent.iot.twcall.databinding.ActivityIpcBinding
import com.tencent.iot.video.device.VideoNativeInterface
import com.tencent.iot.video.device.annotations.CmFrameType
import com.tencent.iot.video.device.annotations.CsChannelType
import com.tencent.iot.video.device.annotations.StreamType
import com.tencent.iot.video.device.callback.IvCsInitCallback
import com.tencent.iot.video.device.model.AvDataInfo
import com.tencent.iot.video.device.model.CsBalanceInfo
import com.tencent.iot.video.device.model.CsChannelInfo
import com.tencent.iot.video.device.model.CsEventResultInfo
import com.tencent.iot.video.device.model.CsNotifyMsgData
import com.tencent.iotvideo.link.CameraRecorder
import com.tencent.iotvideo.link.SimplePlayer
import com.tencent.iotvideo.link.listener.OnEncodeListener
import com.tencent.iotvideo.link.util.copyTextToClipboard
import com.tencent.iotvideo.link.util.updateOperate
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream

private const val TAG = "IPCActivity"

class IPCActivity : BaseIPCActivity<ActivityIpcBinding>(), IvCsInitCallback, OnEncodeListener {

    private val player = SimplePlayer()
    private val cameraRecorder = CameraRecorder()
    private var lastClickTime = 0L
    private var localPreviewSurface: SurfaceTexture? = null
    private val handler = Handler(Looper.getMainLooper())
    private val UPDATE_P2P_INFO_TOKEN = "update_p2p_info_token"
    private val CHECK_PREPARE_CS_TOKEN = "check_prepare_cs_token"
    private var customCommandDialog: CustomCommandDialog? = null
    private var avDataInfo: AvDataInfo? = null
    private var shouldCsInit: Boolean = true
    private var csBalanceInfo: CsBalanceInfo? = null
    private var wareReportSate = false

    private val listener = object : TextureView.SurfaceTextureListener {
        override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
            if (surface == binding.textureViewIpc.surfaceTexture) {
                // Initialize the SurfaceTexture object
                localPreviewSurface = surface

                // Start the camera encoder
                cameraRecorder.openCamera(binding.textureViewIpc, this@IPCActivity)
                Log.d("IPCActivity", "onSurfaceTextureAvailable")
            }
        }

        override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
            // Not used in this example
            Log.d("IPCActivity", "onSurfaceTextureSizeChanged")
        }

        override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
            Log.d("IPCActivity", "onSurfaceTextureDestroyed")
            if (surface == binding.textureViewIpc.surfaceTexture) {
                // Stop the camera encoder
                cameraRecorder.closeCamera()
            }
            return true
        }

        override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {
            // Not used in this example
        }
    }

    override fun getViewBinding(): ActivityIpcBinding = ActivityIpcBinding.inflate(layoutInflater)
    override fun initView() {
        prepareCs()
        cameraRecorder.setOnEncodeListener(this)
        with(binding) {
            titleLayout.tvTitle.text = getString(R.string.title_ipc)
            titleLayout.ivBack.setOnClickListener { onBackPressedDispatcher.onBackPressed() }
            binding.titleLayout.ivRightBtn.isVisible = true
            binding.titleLayout.ivRightBtn.setOnClickListener {
                val dialog = QualitySettingDialog(this@IPCActivity)
                dialog.setOnDismissListener {
                    cameraRecorder.closeCamera()
                    cameraRecorder.openCamera(binding.textureViewIpc, this@IPCActivity)
                }
                dialog.show(supportFragmentManager)
            }
            textDevInfo.text =
                String.format((getString(R.string.text_device_info)), "${productId}_$deviceName")
            textureViewIpc.surfaceTextureListener = listener
            tvCopy.setOnClickListener {
                copyTextToClipboard(this@IPCActivity, tvP2pInfo.text.toString().substringAfter(":"))
                showToast("已复制p2p信息")
            }
            btnCloudStorageReport.setOnClickListener {
                if (!checkCsInfo()) return@setOnClickListener
                VideoNativeInterface.getInstance()
                    .stopCsEvent(CsChannelType.CS_SINGLE_CH, 1, "report test cs info")
                val csEventRes = VideoNativeInterface.getInstance()
                    .startCsEvent(CsChannelType.CS_SINGLE_CH, 1, "report test cs info")
                if (csEventRes != 0) {
                    showToast("触发事件失败,res:$csEventRes")
                } else {
                    showToast("触发事件成功")
                }
            }
            btnCloudPicWareReport.setOnClickListener {
                if (!checkCsInfo()) return@setOnClickListener
                val bitmap = binding.textureViewIpc.getBitmap(
                    cameraRecorder.mVideoWidth,
                    cameraRecorder.mVideoHeight
                )
                if (bitmap != null) {
                    val reportRes = VideoNativeInterface.getInstance().reportCsEventDirectly(
                        CsChannelType.CS_SINGLE_CH,
                        1,
                        "report test cs info",
                        0,
                        bitmapToByteArray(bitmap)
                    )
                    if (reportRes == 0) {
                        showToast("告警事件上报成功")
                    } else {
                        showToast("告警事件上报失败，res:$reportRes")
                    }
                }
            }
            btnCloudVideoWareReport.setOnClickListener {
                if (!checkCsInfo()) return@setOnClickListener
                if (!wareReportSate) {
                    VideoNativeInterface.getInstance().startCsEvent(
                        CsChannelType.CS_SINGLE_CH, 2, "start report test cs video info"
                    )
                } else {
                    VideoNativeInterface.getInstance().stopCsEvent(
                        CsChannelType.CS_SINGLE_CH, 2, "stop report test cs video info"
                    )
                }
            }
            btnCustomCommand.setOnClickListener {
                if (!isOnline) {
                    showToast("设备未上线")
                    return@setOnClickListener
                }
                customCommandDialog = CustomCommandDialog(this@IPCActivity, visitor)
                customCommandDialog?.show(supportFragmentManager)
            }
//            btnIpcCall.setOnClickListener {
//                val time = System.currentTimeMillis()
//                val timeD = time - lastClickTime
//                //防止频繁点击
//                if (timeD in 1..999) {
//                    Toast.makeText(this@IPCActivity, "频繁点击！", Toast.LENGTH_SHORT).show()
//                    return@setOnClickListener
//                }
//                lastClickTime = time
//                // msg_id 6: 按门铃
//                VideoNativeInterface.getInstance().sendMsgNotice(6)
//            }
        }
    }

    private val taskInitCsRunnable = Runnable {
        prepareCs()
    }

    private fun prepareCs() {
        if (isOnline) {
            initCs()
        } else {
            if (Build.VERSION.SDK_INT >= 28) {
                handler.postDelayed(taskInitCsRunnable, CHECK_PREPARE_CS_TOKEN, 1000)
            } else {
                handler.postDelayed(taskInitCsRunnable, 1000)
            }
        }
    }

    private fun checkCsInfo(): Boolean {
        if (!isOnline) {
            showToast("设备未上线")
            return false
        }
        if (csBalanceInfo == null) {
            showToast("设备未获取到云存套餐")
            return false
        } else if (csBalanceInfo?.csSwitch == 0) {
            showToast("设备未开通云存套餐")
            return false
        }
        if (shouldCsInit) {
            showToast("未初始化云存")
            return false
        }
        return true
    }

    private fun initCs(): Int {
        if (shouldCsInit) {
            val csChannelInfo = arrayOfNulls<CsChannelInfo>(1)
            for (i in csChannelInfo.indices) {
                val channelInfo = CsChannelInfo()
                channelInfo.channelId = i
                channelInfo.u32MaxGopSize = 512 * 1024
                channelInfo.csFormat = 0
                if (avDataInfo != null) {
                    channelInfo.avDataInfo = avDataInfo
                } else {
                    channelInfo.avDataInfo = AvDataInfo.createDefaultAvDataInfo(videoResType)
                }
                channelInfo.eventReportOpt = 0
                csChannelInfo[i] = channelInfo
            }
            shouldCsInit = false
            return VideoNativeInterface.getInstance().initCs(csChannelInfo, this)
        }
        return -1
    }

    private fun updateP2pInfo() {
        val p2pInfo = VideoNativeInterface.getInstance().p2pInfo
        if (p2pInfo?.isNotEmpty() == true) {
            if (!binding.tvP2pInfo.text.toString().contains(p2pInfo)) {
                showToast("P2PInfo 已更新")
            }
            binding.tvP2pInfo.text = String.format(getString(R.string.text_p2p_info), p2pInfo)
            if (Build.VERSION.SDK_INT >= 28) {
                handler.postDelayed(taskRunnable, UPDATE_P2P_INFO_TOKEN, 10000)
            } else {
                handler.postDelayed(taskRunnable, 10000)
            }
        }
    }

    private var shouldGetCs = true
    private fun updateCsState() {
        if (shouldGetCs) {
            csBalanceInfo =
                VideoNativeInterface.getInstance().getCsBalanceInfo(CsChannelType.CS_SINGLE_CH, 0)
            if (csBalanceInfo != null) {
                shouldGetCs = false
                binding.tvCloudStorageState.text = String.format(
                    getString(R.string.text_cloud_storage_state),
                    if (csBalanceInfo?.csSwitch == 1) {
                        binding.btnCloudPicWareReport.updateOperate(true)
                        binding.btnCloudVideoWareReport.updateOperate(true)
                        binding.btnCloudStorageReport.updateOperate(true)
                        "已开通"
                    } else "未开通"
                )
            }
        }
    }

    private val taskRunnable = Runnable {
        updateP2pInfo()
        updateCsState()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (Build.VERSION.SDK_INT >= 28) {
            handler.removeCallbacksAndMessages(UPDATE_P2P_INFO_TOKEN)
        } else {
            handler.removeCallbacks(taskRunnable)
        }
        if (Build.VERSION.SDK_INT >= 28) {
            handler.removeCallbacksAndMessages(CHECK_PREPARE_CS_TOKEN)
        } else {
            handler.removeCallbacks(taskInitCsRunnable)
        }
        if (!shouldCsInit) {
            VideoNativeInterface.getInstance().exitCs()
        }
    }

    override fun onOnline(netDateTime: Long) {
        super.onOnline(netDateTime)
        lifecycleScope.launch(Dispatchers.Main) {
            updateP2pInfo()
            if (Build.VERSION.SDK_INT >= 28) {
                handler.postDelayed(taskRunnable, UPDATE_P2P_INFO_TOKEN, 5000)
            } else {
                handler.postDelayed(taskRunnable, 5000)
            }
            binding.btnCustomCommand.updateOperate(true)
        }
    }

    override fun onGetAvEncInfo(visitor: Int, channel: Int, videoResType: Int): AvDataInfo {
        avDataInfo = AvDataInfo.createDefaultAvDataInfo(videoResType)
        return avDataInfo!!
    }

    override fun onStartRealPlay(visitor: Int, channel: Int, videoResType: Int) {
        super.onStartRealPlay(visitor, channel, videoResType)
        cameraRecorder.startRecording(visitor, channel, videoResType)
    }

    override fun onStopRealPlay(visitor: Int, channel: Int, videoResType: Int) {
        cameraRecorder.stopRecording(visitor, videoResType)
    }

    override fun onStartRecvAudioStream(
        visitor: Int,
        channel: Int,
        type: Int,
        option: Int,
        mode: Int,
        width: Int,
        sample_rate: Int,
        sample_num: Int
    ): Int {
        return player.startAudioPlay(visitor, type, option, mode, width, sample_rate, sample_num)
    }

    override fun onStartRecvVideoStream(
        visitor: Int,
        channel: Int,
        type: Int,
        height: Int,
        width: Int,
        frameRate: Int
    ): Int {
        return super.onStartRecvVideoStream(visitor, channel, type, height, width, frameRate)
    }

    override fun onStopRecvStream(visitor: Int, channel: Int, streamType: Int): Int {
        return if (streamType == StreamType.IV_AVT_STREAM_TYPE_VIDEO) {
            player.stopVideoPlay(visitor)
        } else {
            player.stopAudioPlay(visitor)
        }
    }

    override fun onRecvStream(
        visitor: Int,
        streamType: Int,
        data: ByteArray?,
        len: Int,
        pts: Long,
        seq: Long
    ): Int {
        if (streamType == 1) {
            return player.playVideoStream(visitor, data, len, pts, seq)
        } else if (streamType == 0) {
            return player.playAudioStream(visitor, data, len, pts, seq)
        }
        return 0
    }

    override fun onRecvCommand(
        command: Int,
        visitor: Int,
        channel: Int,
        videoResType: Int,
        args: String?
    ): String {
        return customCommandDialog?.receiveCommand(args ?: "null").toString()
    }

    private fun drawableToBitmap(drawable: Drawable): Bitmap {
        if (drawable is BitmapDrawable) {
            return drawable.bitmap
        }

        val width = drawable.intrinsicWidth
        val height = drawable.intrinsicHeight
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

    // 将Bitmap转换为byte数组
    private fun bitmapToByteArray(bitmap: Bitmap): ByteArray {
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
        return stream.toByteArray()
    }

    override fun onStartPushStream(channel: Int): Int {
        Log.d(TAG, "onStartPushStream  channel:$channel")
        wareReportSate = true
        lifecycleScope.launch(Dispatchers.Main) {
            binding.btnCloudVideoWareReport.text = "正在上报告警中..."
        }
        return 0
    }

    override fun onStopPushStream(channel: Int): Int {
        Log.d(TAG, "onStopPushStream  channel:$channel")
        wareReportSate = false
        lifecycleScope.launch(Dispatchers.Main) {
            binding.btnCloudVideoWareReport.text = getString(R.string.text_cloud_video_ware)
        }
        return 0
    }

    override fun onAiServiceNotify(channel: Int, aiServerType: Int, utcExpire: Long) {
        Log.d(TAG, "onAiServiceNotify  channel:$channel")
    }

    override fun onEventCapturePicture(channel: Int, eventId: Int): ByteArray {
        Log.d(TAG, "onEventCapturePicture  channel:$channel")
        val bitmap = binding.textureViewIpc.getBitmap(
            cameraRecorder.mVideoWidth,
            cameraRecorder.mVideoHeight
        )
        if (bitmap != null) {
            return bitmapToByteArray(bitmap)
        }
        return ByteArray(0)
    }

    override fun onEventPictureResult(channel: Int, errCode: Int): Int {
        Log.d(TAG, "onEventPictureResult  channel:$channel")
        return 0
    }

    override fun onEventReportResult(channel: Int, resultInfo: CsEventResultInfo?): Int {
        Log.d(TAG, "onEventReportResult  channel:$channel")
        return 0
    }

    override fun onNotify(channel: Int, notifyMsgType: Int, notifyData: CsNotifyMsgData?): Int {
        Log.d(TAG, "onNotify  channel:$channel")
        return 0
    }

    override fun onGetBalance(
        channel: Int,
        isValid: Boolean,
        balanceInfo: CsBalanceInfo?,
        timeoutMs: Int
    ) {
        Log.d(TAG, "onGetBalance  channel:$channel")
    }

    override fun onDumpFile(
        channel: Int,
        state: Int,
        startTs: Long,
        endTs: Long,
        fileName: String?,
        buf: ByteArray?,
        len: Int
    ): Int {
        Log.d(TAG, "onDumpFile  channel:$channel")
        return 0
    }

    override fun onAudioEncoded(datas: ByteArray?, pts: Long, seq: Long) {
        if (wareReportSate) {
            VideoNativeInterface.getInstance().pushCsAudioStream(
                CsChannelType.CS_SINGLE_CH, datas, pts, seq.toInt()
            )
        }
    }

    override fun onVideoEncoded(datas: ByteArray?, pts: Long, seq: Long, isKeyFrame: Boolean) {
        if (wareReportSate) {
            val type = if (isKeyFrame) {
                CmFrameType.IV_CM_FRAME_TYPE_I
            } else {
                CmFrameType.IV_CM_FRAME_TYPE_P
            }
            VideoNativeInterface.getInstance().pushCsVideoStream(
                CsChannelType.CS_SINGLE_CH, datas, pts, type, seq.toInt()
            )
        }
    }
}